assembl/views/api2/votes.py

```diff
@@ -349,3 +350,69 @@ def global_vote_results_csv(request):
         csvw.writerow(row)
     output.seek(0)
     return Response(body_file=output, content_type='text/csv', content_disposition='attachment; filename="vote_results.csv')
+
+
+@view_config(context=InstanceContext, name="extract_csv_voters",
+             ctx_instance_class=VotingWidget, request_method='GET',
+             permission=P_DISC_STATS)
+def extract_voters(request):
+    import assembl.models as m
+    import numpy as np
+    extract_list = []
+    ctx = request.context
+    widget = ctx._instance
+    if widget.activity_state != "ended":
+        permissions = get_permissions(user_id, ctx.get_discussion_id())
+        if P_ADMIN_DISC not in permissions:
+            raise HTTPUnauthorized()
+    user_prefs = LanguagePreferenceCollection.getCurrent()
+    db = widget.db
+    fieldnames = ["Nom du contributeur", "Adresse mail du contributeur", "Date/heure du vote", "Proposition"]
+    ideas = widget.db.query(Idea).join(AbstractVoteSpecification).filter(AbstractVoteSpecification.widget_id == widget.id).distinct().all()
+    votes = widget.db.query(AbstractIdeaVote).filter(AbstractVoteSpecification.widget_id==widget.id).all()
+    votes.sort(key=lambda x: x.vote_spec_id, reverse=True)
+    users = widget.db.query(User).all()
+    vote_specs = widget.db.query(AbstractVoteSpecification).filter(AbstractVoteSpecification.widget_id == widget.id).all()
+
+    for count,vote in enumerate(votes):
+        voter = m.User.get(vote.voter_id)
+        contributor = voter.name or ""
+        contributor_mail = voter.preferred_email or ""
+        vote_date = vote.vote_date or ""
+        proposition = m.Idea.get(vote.idea_id).title.best_lang(user_prefs).value or ""
+        vote_value = vote.vote_value
+
+        if votes[count].vote_spec_id != votes[count-1].vote_spec_id:
+            vote_spec = m.AbstractVoteSpecification.get(votes[count].vote_spec_id)
+            fieldnames.append("  ")
+
+        extract_info = {
+            "Nom du contributeur": contributor.encode('utf-8'),
+            "Adresse mail du contributeur": contributor_mail.encode('utf-8'),
+            "Date/heure du vote": str(vote_date).encode('utf-8'),
+            "Proposition": proposition.encode('utf-8'),
+        }
+
+        if vote.type == u'token_idea_vote':
+            token_category = m.TokenCategorySpecification.get(vote.token_category_id).name.best_lang(user_prefs).value or ""
+            if token_category not in fieldnames:
+                fieldnames.append(token_category.encode('utf-8'))
```

Comment:
Hmm, adding columns at runtime will make a debate with multiple votes very hard to read. If you're not adding to the column after this, then fine. It's ugly to read, and if there are multiple token votes with the same category name, it will be very confusing, but for a very primitive first pass, it's okay. These should not be CSVs, instead, they should be excel format, a token specification per sheet.

Reply:
@ayyazdaniaryan Actually, I raised the ugliness of the CSV with Fred, who raised the issue with christy. They want only one CSV no matter how long it is. the only thing that we modified in the user story is to put blank columns between each vote (whether it is token or gauge) , which I did. During Sprint planning, we agreed not to make several sheets per export, unless you know a way to do it. so it was finally decided to to make two seperate buttons.